pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp

```diff
@@ -371,86 +355,249 @@ void applyGenMultiRZ(
         });
 }
 
+template <class PrecisionT, class FuncT> class applyNCGenerator1Functor {
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+    std::size_t mask{0U};
+    static constexpr std::size_t one{1U};
+    std::size_t i0;
+    std::size_t i1;
+
+  public:
+    template <class ExecutionSpace>
+    applyNCGenerator1Functor([[maybe_unused]] ExecutionSpace exec,
+                             Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                             std::size_t num_qubits,
+                             const std::vector<std::size_t> &controlled_wires,
+                             const std::vector<bool> &controlled_values,
+                             const std::vector<std::size_t> &wires,
+                             FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+        const std::size_t n_contr = controlled_wires.size();
+        const std::size_t n_wires = wires.size();
+        const std::size_t nw_tot = n_contr + n_wires;
+        PL_ASSERT(n_wires == 1);
+        PL_ASSERT(num_qubits >= nw_tot);
+
+        std::vector<std::size_t> all_wires;
+        all_wires.reserve(nw_tot);
+        all_wires.insert(all_wires.begin(), wires.begin(), wires.end());
+        all_wires.insert(all_wires.begin(), controlled_wires.begin(),
+                         controlled_wires.end());
+
+        const auto &[parity_, rev_wires_] =
+            reverseWires(num_qubits, wires, controlled_wires);
+        parity = parity_;
+        std::vector<std::size_t> indices_ =
+            generateBitPatterns(all_wires, num_qubits);
+
+        std::size_t k = 0;
+        mask = std::accumulate(
+            controlled_values.rbegin(), controlled_values.rend(),
+            std::size_t{0}, [&k](std::size_t acc, std::size_t value) {
+                return acc | (static_cast<std::size_t>(value) << k++);
+            });
```

Comment:
Beautiful!

Reply:
Thanks Amintor for your good suggestion ;)